heliopy/data/mms.py

```diff
@@ -7,131 +7,1773 @@
+        instr : str, list
+            Instrument ID(s)
+        mode : str, list
+            Data rate mode(s). Options include slow, fast, srvy, brst
+        level : str, list
+            Data level(s). Options include l1a, l1b, l2pre, l2, l3
+        tstart : str, list
+            Start time of data file. In general, the format is
+            YYYYMMDDhhmmss for "brst" mode and YYYYMMDD for "srvy"
+            mode (though there are exceptions). If not given, the
+            default is "*".
+        version : str, list
+            File version, formatted as "X.Y.Z", where X, Y, and Z
+            are integer version numbers.
+        optdesc : str, list
+            Optional file name descriptor. If multiple parts,
+            they should be separated by hyphens ("-"), not under-
+            scores ("_").
+
+    Returns
+    -------
+        fnames : str, list
+            File names constructed from inputs.
+    '''
+
+    # Convert all to lists
+    if isinstance(sc, str):
+        sc = [sc]
+    if isinstance(instr, str):
+        instr = [instr]
+    if isinstance(mode, str):
+        mode = [mode]
+    if isinstance(level, str):
+        level = [level]
+    if isinstance(tstart, str):
+        tstart = [tstart]
+    if isinstance(version, str):
+        version = [version]
+    if optdesc is not None and isinstance(optdesc, str):
+        optdesc = [optdesc]
+
+    # Accept tuples, as those returned by Construct_Filename
+    if type(sc) == 'tuple':
+        sc_ids = [file[0] for file in sc]
+        instr = [file[1] for file in sc]
+        mode = [file[2] for file in sc]
+        level = [file[3] for file in sc]
+        tstart = [file[-2] for file in sc]
+        version = [file[-1] for file in sc]
+
+        if len(sc) > 6:
+            optdesc = [file[4] for file in sc]
+        else:
+            optdesc = None
+    else:
+        sc_ids = sc
+
+    if optdesc is None:
+        fnames = ['_'.join((s, i, m, l, t, 'v'+v+'.cdf'))
+                  for s in sc_ids
+                  for i in instr
+                  for m in mode
+                  for l in level
+                  for t in tstart
+                  for v in version
+                  ]
+    else:
+        fnames = ['_'.join((s, i, m, l, o, t, 'v'+v+'.cdf'))
+                  for s in sc_ids
+                  for i in instr
+                  for m in mode
+                  for l in level
+                  for o in optdesc
+                  for t in tstart
+                  for v in version
+                  ]
+    return fnames
+
+
+def construct_path(*args, data_type='science', **kwargs):
+    '''
+    Construct a directory structure compliant with MMS path guidelines.
+
+    MMS paths follow the convention
+        selections: sitl/type_selections_[gls_type_]
+        brst: sc/instr/mode/level[/optdesc]/<year>/<month>/<day>
+        srvy: sc/instr/mode/level[/optdesc]/<year>/<month>
+
+    Parameters
+    ----------
+        args : dict
```

Comment:
This is one too many indents (`args` should be level with the tile in the previous line)